tracing/optimisation/utils.py

```diff
@@ -0,0 +1,66 @@
+from dataclasses import dataclass
+from traceback import linecache # type: ignore
+import functools
+import io
+import typing
+import tokenize
+
+
+# Start with characters or underscore, rest is the same plus numbers
+@dataclass
+class FrameWithMetadata:
```

Comment:
There aren't any tests to validate the functions. Create a test in which a function with loops is traced and use the frames to generate the meta data, and validate the function return values.